File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/importer/scmlogparser/parser/LogLineParser.java
package de.maibornwolff.codecharta.importer.scmlogparser.parser;

import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit;
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification;
import de.maibornwolff.codecharta.importer.scmlogparser.input.VersionControlledFile;
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.stream.Stream;

/**
 * Parses log lines and generates VersionControlledFiles from them using specific parserStrategy and metricsFactory
 */
public class LogLineParser {
    private final LogParserStrategy parserStrategy;
    private final MetricsFactory metricsFactory;

    public LogLineParser(LogParserStrategy parserStrategy, MetricsFactory metricsFactory) {
        this.parserStrategy = parserStrategy;
        this.metricsFactory = metricsFactory;
    }

    public List<VersionControlledFile> parse(Stream<String> logLines) {
        return logLines.collect(parserStrategy.createLogLineCollector())
                .map(this::parseCommit)
                .collect(CommitCollector.create(metricsFactory));
    }


    Commit parseCommit(List<String> commitLines) {
        String author = parserStrategy.parseAuthor(commitLines).orElseThrow(() -> new IllegalArgumentException("No author found in input"));
        OffsetDateTime commitDate = parserStrategy.parseDate(commitLines).orElseThrow(() -> new IllegalArgumentException("No commit date found in input"));
        List<Modification> modifications = parserStrategy.parseModifications(commitLines);
        return new Commit(author, modifications, commitDate);
    }
}

File: analysis/import/SCMLogParser/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparser/parser/LogLineParser.kt
package de.maibornwolff.codecharta.importer.scmlogparser.parser

import de.maibornwolff.codecharta.ProgressTracker
import de.maibornwolff.codecharta.importer.scmlogparser.input.Commit
import de.maibornwolff.codecharta.importer.scmlogparser.input.Modification
import de.maibornwolff.codecharta.importer.scmlogparser.input.VersionControlledFile
import de.maibornwolff.codecharta.importer.scmlogparser.input.metrics.MetricsFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.time.OffsetDateTime
import java.util.stream.Stream

/**
 * Parses log lines and generates VersionControlledFiles from them using specific parserStrategy and metricsFactory
 */
class LogLineParser(
    private val parserStrategy: LogParserStrategy,
    private val metricsFactory: MetricsFactory,
    private val silent: Boolean = false,
    private val logSizeInByte: Long = 0
) {

    private var currentBytesParsed = 0L
    private val progressTracker: ProgressTracker = ProgressTracker()
    private val parsingUnit = "Byte"

    fun parse(logLines: Stream<String>): List<VersionControlledFile> {
        val parsedCommit = logLines.collect(parserStrategy.createLogLineCollector())
            .map { this.parseCommit(it) }.filter { !it.isEmpty }
            .collect(CommitCollector.create(metricsFactory))

        progressTracker.updateProgress(logSizeInByte, logSizeInByte, parsingUnit)

        return parsedCommit
    }

    internal fun parseCommit(commitLines: List<String>): Commit {
        return try {
            var author = ""
            var commitDate = OffsetDateTime.now()
            var modifications: List<Modification> = listOf()

            runBlocking(Dispatchers.Default) {
                launch {
                    author = parserStrategy.parseAuthor(commitLines)
                    commitDate = parserStrategy.parseDate(commitLines)
                    modifications = parserStrategy.parseModifications(commitLines)
                }
            }

            commitLines.forEach {
                currentBytesParsed += it.length
            }

            if (!silent) progressTracker.updateProgress(logSizeInByte, currentBytesParsed, parsingUnit)
            Commit(author, modifications, commitDate)
        } catch (e: NoSuchElementException) {
            System.err.println("Skipped commit with invalid syntax ($commitLines)")
            Commit("", listOf(), OffsetDateTime.now())
        }
    }
}
